controllers/component/component_adapter.go

```diff
@@ -0,0 +1,160 @@
+	"github.com/redhat-appstudio/integration-service/metrics"
+	"github.com/redhat-appstudio/operator-toolkit/controller"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// Adapter holds the objects needed to reconcile a integration PipelineRun.
+type Adapter struct {
+	component   *applicationapiv1alpha1.Component
+	application *applicationapiv1alpha1.Application
+	loader      loader.ObjectLoader
+	logger      h.IntegrationLogger
+	client      client.Client
+	context     context.Context
+}
+
+// NewAdapter creates and returns an Adapter instance.
+func NewAdapter(component *applicationapiv1alpha1.Component, application *applicationapiv1alpha1.Application, logger h.IntegrationLogger, loader loader.ObjectLoader, client client.Client,
+	context context.Context) *Adapter {
+	return &Adapter{
+		component:   component,
+		application: application,
+		logger:      logger,
+		loader:      loader,
+		client:      client,
+		context:     context,
+	}
+}
+
+// EnsureComponentIsCleanedUp is an operation that will ensure components
+// marked for deletion have a snapshot created without said component
+func (a *Adapter) EnsureComponentIsCleanedUp() (controller.OperationResult, error) {
+	if !hasComponentBeenDeleted(a.component) {
+		return controller.ContinueProcessing()
+	}
+
+	applicationComponents, err := a.loader.GetAllApplicationComponents(a.client, a.context, a.application)
+	if err != nil {
+		a.logger.Error(err, "Failed to load application components")
+		return controller.RequeueWithError(err)
+	}
+
+	var snapshotComponents []applicationapiv1alpha1.SnapshotComponent
+
+	for _, individualComponent := range *applicationComponents {
+		component := individualComponent
+		if a.component.Name != component.Name {
+			containerImage := component.Spec.ContainerImage
+			componentSource := gitops.GetComponentSourceFromComponent(&component)
+			snapshotComponents = append(snapshotComponents, applicationapiv1alpha1.SnapshotComponent{
+				Name:           component.Name,
+				ContainerImage: containerImage,
+				Source:         *componentSource,
+			})
+		}
+	}
+
+	err = a.failStaleSnapshots()
+	if err != nil {
+		a.logger.Error(err, "Failed to update stale snapshots")
+		return controller.StopProcessing()
+	}
+
+	if len(snapshotComponents) == 0 {
+		a.logger.Info("Application has no available snapshot components for snapshot creation")
+		return controller.StopProcessing()
+	}
+
+	_, err = a.createUpdatedSnapshot(&snapshotComponents)
+	if err != nil {
+		a.logger.Error(err, "Failed to create new snapshot after component deletion")
+		return controller.RequeueWithError(err)
+	}
+
+	return controller.ContinueProcessing()
+}
+
+func (a *Adapter) failStaleSnapshots() error {
+	allSnapshots, err := a.loader.GetAllSnapshots(a.client, a.context, a.application)
+	if err != nil {
+		a.logger.Error(err, "Failed to load all snapshots")
+		return err
+	}
+	for _, snapshot := range *allSnapshots {
```

Comment:
There is a `G601 (CWE-118): Implicit memory aliasing in for loop.` Go Security [CI failure](https://github.com/redhat-appstudio/integration-service/actions/runs/6307716804/job/17124768810?pr=273) for this loop.
It can be resolved pretty simply by adding an assignment for the `snapshot` variable under this line like e.g. [here](https://github.com/redhat-appstudio/integration-service/blob/02c829368d912e684b82dbab204b34f56e576fba/controllers/snapshot/snapshot_adapter.go#L106).